hi-ml/src/health_ml/utils/regression_test_utils.py

```diff
@@ -0,0 +1,169 @@
+    based on the extension of this file.
+    :param actual: A file to be checked against the expected file.
+    :param csv_relative_tolerance: When comparing CSV files, maximum allowed relative discrepancy.
+    If 0.0, do not allow any discrepancy.
+    :return: An empty string if the files are identical, an error message with details otherwise.
+    """
+    def print_lines(prefix: str, lines: List[str]) -> None:
+        num_lines = len(lines)
+        count = min(5, num_lines)
+        logging.info(f"{prefix} {num_lines} lines, first {count} of those:")
+        logging.info(os.linesep.join(lines[:count]))
+
+    def try_read_csv(prefix: str, file: Path) -> Optional[pd.DataFrame]:
+        try:
+            return pd.read_csv(file)
+        except Exception as ex:
+            logging.info(f"{prefix} file can't be read as CSV: {str(ex)}")
+            return None
+
+    if expected.suffix == CSV_SUFFIX:
+        expected_df = try_read_csv("Expected", expected)
+        actual_df = try_read_csv("Actual", actual)
+        if expected_df is None or actual_df is None:
+            return FILE_FORMAT_ERROR
+        try:
+            pd.testing.assert_frame_equal(actual_df, expected_df, rtol=csv_relative_tolerance)
+        except Exception as ex:
+            logging.info(str(ex))
+            return CONTENTS_MISMATCH
+    elif expected.suffix in TEXT_FILE_SUFFIXES:
+        # Compare line-by-line to avoid issues with line separators
+        expected_lines = expected.read_text().splitlines()
+        actual_lines = actual.read_text().splitlines()
+        if expected_lines != actual_lines:
+            print_lines("Expected", expected_lines)
+            print_lines("Actual", actual_lines)
+            return CONTENTS_MISMATCH
+    else:
+        expected_binary = expected.read_bytes()
+        actual_binary = actual.read_bytes()
+        if expected_binary != actual_binary:
+            logging.info(f"Expected {len(expected_binary)} bytes, actual {len(actual_binary)} bytes")
+            return CONTENTS_MISMATCH
+    return ""
+
+
+def compare_folder_contents(expected_folder: Path,
+                            csv_relative_tolerance: float,
+                            actual_folder: Optional[Path] = None,
+                            run: Optional[Run] = None) -> List[str]:
+    """
+    Compares a set of files in a folder, against files in either the other folder or files stored in the given
+    AzureML run. Each file that is present in the "expected" folder must be also present in the "actual" folder
+    (or the AzureML run), with exactly the same contents, in the same folder structure.
+    For example, if there is a file "<expected>/foo/bar/contents.txt", then there must also be a file
+    "<actual>/foo/bar/contents.txt"
+
+    :param expected_folder: A folder with files that are expected to be present.
+    :param actual_folder: The output folder with the actually produced files.
+    :param run: An AzureML run
```

Comment:
Can we elaborate on this? Should only `run` or `actual_folder` be passed? What happens if both or neither are passed?

Reply:
done